### PR TITLE
TST: special.logsumexp: modernise tests

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
n/a, precursor to reviving gh-20935

#### What does this implement/fix?
Basic refactor into a class and removing use of `np.testing.assert_`

### CODE REVIEW

File: scipy/special/tests/test_logsumexp.py [hunk 1]
```diff
@@ -1,212 +1,209 @@
...
 
-def test_logsumexp_sign_shape():
-    a = np.ones((1,2,3,4))
-    b = np.ones_like(a)
+    def test_logsumexp_sign_shape(self):
+        a = np.ones((1,2,3,4))
+        b = np.ones_like(a)
+
+        r, s = logsumexp(a, axis=2, b=b, return_sign=True)
+        assert r.shape == s.shape == (1,2,4)
```
Comment (mdhaber) [inline scipy/special/tests/test_logsumexp.py:114]:
```suggestion
        assert r.shape == s.shape == (1, 2, 4)
```

File: scipy/special/tests/test_logsumexp.py [hunk 2]
```diff
@@ -1,212 +1,209 @@
...
 
-    r, s = logsumexp(a, axis=2, b=b, return_sign=True)
+        r, s = logsumexp(a, axis=(1,3), b=b, return_sign=True)
+        assert r.shape == s.shape == (1,3)
```
Comment (mdhaber) [inline scipy/special/tests/test_logsumexp.py:117]:
```suggestion
        assert r.shape == s.shape == (1, 3)
```

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
Approved with PEP8 updates, but should the `softmax` tests be a separate class?

**Comment by mdhaber** [comment]
Not at this time, but I'd be happy to review one.

**Review by j-bowhay — APPROVED** [review]
Thanks, there are a few places the indentation is off but converting to array API will change this anyway so no need to fix it here

**Pull request merged**